pkg/server/cache/entrycache/fullcache_test.go

```diff
@@ -0,0 +1,787 @@
+			SpiffeId:  "spiffe://example.org/workload2",
+			Selectors: irrelevantSelectors,
+		},
+		{
+			ParentId:  agentIDs[2].String(),
+			SpiffeId:  "spiffe://example.org/workload3",
+			Selectors: irrelevantSelectors,
+		},
+	}
+
+	workloadEntries := make([]*common.RegistrationEntry, len(workloadEntriesToCreate))
+	for i, e := range workloadEntriesToCreate {
+		workloadEntries[i] = createRegistrationEntry(ctx, t, ds, e)
+	}
+
+	for i, agentID := range agentIDs {
+		node := &common.AttestedNode{
+			SpiffeId:            agentID.String(),
+			AttestationDataType: testNodeAttestor,
+			CertSerialNumber:    strconv.Itoa(i),
+			CertNotAfter:        time.Now().Add(24 * time.Hour).Unix(),
+		}
+
+		createAttestedNode(t, ds, node)
+	}
+
+	setNodeSelectors(ctx, t, ds, agentIDs[0].String(), s1, s2)
+	setNodeSelectors(ctx, t, ds, agentIDs[1].String(), s1, s3)
+
+	cache, err := BuildFromDataStore(context.Background(), ds)
+	assert.NoError(t, err)
+
+	assertAuthorizedEntries := func(agentID spiffeid.ID, entries ...*common.RegistrationEntry) {
+		expected, err := api.RegistrationEntriesToProto(entries)
+		require.NoError(t, err)
+		assert.ElementsMatch(t, expected, cache.GetAuthorizedEntries(agentID))
+	}
+
+	assertAuthorizedEntries(agentIDs[0], append(nodeAliasEntries, workloadEntries[:2]...)...)
+	assertAuthorizedEntries(agentIDs[1], nodeAliasEntries[1], workloadEntries[1])
+	assertAuthorizedEntries(agentIDs[2], workloadEntries[2])
+}
+
+func TestFullCacheExcludesNodeSelectorMappedEntriesForExpiredAgents(t *testing.T) {
+	// This test verifies that the cache contains no workloads parented to alias entries
+	// that are only associated with an expired agent.
+	//
+	// Data used in this test:
+	//
+	// Registration entry graph:
+	// (agent SPIFFE IDs are shown as parented to the root for simplicity of illustrating the hierarchy)
+	//
+	//           ---------------------------root------------------------
+	//          /             |              |               |          \
+	//   group/0          group/1         group/2      agent/active    agent/expired
+	//      |                |              |                |            \
+	//  workload/0       workload/1     workload/2      workload/3     workload/4
+	//
+	// Agents:
+	// - agent/active - has a CertNotAfter that is still valid
+	// - agent/expired - has a CertNotAfter that expired
+	//
+	// agent/active maps to group/0 and group/1 based on selector subset matches.
+	// agent/expired maps to group/0 and group/2 based on selector subset matches.
+	//
+	// Normally, agent/expired should be authorized to receive group/0, workload/0, group/2, workload/2, and workload/4.
+	// However, the cache filters out all entries related to the expired agent other than ones shared with other Agents
+	// through node selector subset matching - in this case, just workload/0.
+	// In reality, an expired agent should not be able to request its authorized entries because endpoint security (mTLS)
```

Comment:
nit: mTLS will only verify the SVID validity during the handshake. Since agent connections live for several minutes, the authorization middleware takes care of ensuring that the SVID is still valid on each RPC request over the "long-lived" connection.

Reply:
That's true. I updated the wording to be more specific in that regard.